diff --git a/hotspot/src/cpu/aarch64/vm/assembler_aarch64.hpp b/hotspot/src/cpu/aarch64/vm/assembler_aarch64.hpp
index 8df0c512..f0cf9a4f 100644
--- a/hotspot/src/cpu/aarch64/vm/assembler_aarch64.hpp
+++ b/hotspot/src/cpu/aarch64/vm/assembler_aarch64.hpp
@@ -27,6 +27,7 @@
 #define CPU_AARCH64_VM_ASSEMBLER_AARCH64_HPP
 
 #include "asm/register.hpp"
+#include "runtime/os.hpp"
 
 // definitions of various symbolic names for machine registers
 
@@ -205,6 +206,7 @@ public:
   }
 
   static void patch(address a, int msb, int lsb, uint64_t val) {
+    a = mirror_w(a);
     int nbits = msb - lsb + 1;
     guarantee(val < (1U << nbits), "Field too big for insn");
     assert_cond(msb >= lsb);
@@ -218,6 +220,7 @@ public:
   }
 
   static void spatch(address a, int msb, int lsb, int64_t val) {
+    a = mirror_w(a);
     int nbits = msb - lsb + 1;
     int64_t chk = val >> (nbits - 1);
     guarantee (chk == -1 || chk == 0, "Field too big for insn");
diff --git a/hotspot/src/cpu/aarch64/vm/bytes_aarch64.hpp b/hotspot/src/cpu/aarch64/vm/bytes_aarch64.hpp
index 9d0d5f8a..9cf017c0 100644
--- a/hotspot/src/cpu/aarch64/vm/bytes_aarch64.hpp
+++ b/hotspot/src/cpu/aarch64/vm/bytes_aarch64.hpp
@@ -28,6 +28,7 @@
 #define CPU_AARCH64_VM_BYTES_AARCH64_HPP
 
 #include "memory/allocation.hpp"
+#include "runtime/os.hpp"
 
 class Bytes: AllStatic {
  public:
@@ -43,9 +44,9 @@ class Bytes: AllStatic {
   static inline u4   get_native_u4(address p)         { return *(u4*)p; }
   static inline u8   get_native_u8(address p)         { return *(u8*)p; }
 
-  static inline void put_native_u2(address p, u2 x)   { *(u2*)p = x; }
-  static inline void put_native_u4(address p, u4 x)   { *(u4*)p = x; }
-  static inline void put_native_u8(address p, u8 x)   { *(u8*)p = x; }
+  static inline void put_native_u2(address p, u2 x)   { *(u2*)mirror_w(p) = x; }
+  static inline void put_native_u4(address p, u4 x)   { *(u4*)mirror_w(p) = x; }
+  static inline void put_native_u8(address p, u8 x)   { *(u8*)mirror_w(p) = x; }
 
 
   // Efficient reading and writing of unaligned unsigned data in Java
diff --git a/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.cpp b/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.cpp
index 4713d83f..415d2050 100644
--- a/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.cpp
+++ b/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.cpp
@@ -132,7 +132,7 @@ intptr_t NativeMovConstReg::data() const {
 void NativeMovConstReg::set_data(intptr_t x) {
   if (maybe_cpool_ref(instruction_address())) {
     address addr = MacroAssembler::target_addr_for_insn(instruction_address());
-    *(intptr_t*)addr = x;
+    *(intptr_t*)mirror_w(addr) = x;
   } else {
     MacroAssembler::pd_patch_instruction(instruction_address(), (address)x);
     ICache::invalidate_range(instruction_address(), instruction_size);
@@ -164,7 +164,7 @@ void NativeMovRegMem::set_offset(int x) {
   unsigned insn = *(unsigned*)pc;
   if (maybe_cpool_ref(pc)) {
     address addr = MacroAssembler::target_addr_for_insn(pc);
-    *(int64_t*)addr = x;
+    *(int64_t*)mirror_w(addr) = x;
   } else {
     MacroAssembler::pd_patch_instruction(pc, (address)intptr_t(x));
     ICache::invalidate_range(instruction_address(), instruction_size);
@@ -263,7 +263,7 @@ bool NativeInstruction::is_sigill_zombie_not_entrant() {
 }
 
 void NativeIllegalInstruction::insert(address code_pos) {
-  *(juint*)code_pos = 0xd4bbd5a1; // dcps1 #0xdead
+  *(juint*)mirror_w(code_pos) = 0xd4bbd5a1; // dcps1 #0xdead
 }
 
 //-------------------------------------------------------------------
@@ -284,7 +284,7 @@ void NativeJump::patch_verified_entry(address entry, address verified_entry, add
     guarantee(disp < 1 << 27 && disp > - (1 << 27), "branch overflow");
 
     unsigned int insn = (0b000101 << 26) | ((disp >> 2) & 0x3ffffff);
-    *(unsigned int*)verified_entry = insn;
+    *(unsigned int*)mirror_w(verified_entry) = insn;
   } else {
     // We use an illegal instruction for marking a method as
     // not_entrant or zombie.
diff --git a/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.hpp b/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.hpp
index 7f847edd..a682cf93 100644
--- a/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.hpp
+++ b/hotspot/src/cpu/aarch64/vm/nativeInst_aarch64.hpp
@@ -71,6 +71,7 @@ class NativeInstruction VALUE_OBJ_CLASS_SPEC {
 
  protected:
   address addr_at(int offset) const    { return address(this) + offset; }
+  address addr_rw_at(int offset) const { return mirror_w(addr_at(offset)); }
 
   s_char sbyte_at(int offset) const    { return *(s_char*) addr_at(offset); }
   u_char ubyte_at(int offset) const    { return *(u_char*) addr_at(offset); }
@@ -83,11 +84,11 @@ class NativeInstruction VALUE_OBJ_CLASS_SPEC {
   oop  oop_at (int offset) const       { return *(oop*) addr_at(offset); }
 
 
-  void set_char_at(int offset, char c)        { *addr_at(offset) = (u_char)c; }
-  void set_int_at(int offset, jint  i)        { *(jint*)addr_at(offset) = i; }
-  void set_uint_at(int offset, jint  i)       { *(juint*)addr_at(offset) = i; }
-  void set_ptr_at (int offset, address  ptr)  { *(address*) addr_at(offset) = ptr; }
-  void set_oop_at (int offset, oop  o)        { *(oop*) addr_at(offset) = o; }
+  void set_char_at(int offset, char c)        { *addr_rw_at(offset) = (u_char)c; }
+  void set_int_at(int offset, jint  i)        { *(jint*)addr_rw_at(offset) = i; }
+  void set_uint_at(int offset, jint  i)       { *(juint*)addr_rw_at(offset) = i; }
+  void set_ptr_at (int offset, address  ptr)  { *(address*)addr_rw_at(offset) = ptr; }
+  void set_oop_at (int offset, oop  o)        { *(oop*)addr_rw_at(offset) = o; }
 
  public:
 
diff --git a/hotspot/src/cpu/aarch64/vm/vtableStubs_aarch64.cpp b/hotspot/src/cpu/aarch64/vm/vtableStubs_aarch64.cpp
index 59d5b04d..b7dbb0ac 100644
--- a/hotspot/src/cpu/aarch64/vm/vtableStubs_aarch64.cpp
+++ b/hotspot/src/cpu/aarch64/vm/vtableStubs_aarch64.cpp
@@ -57,6 +57,7 @@ VtableStub* VtableStubs::create_vtable_stub(int vtable_index) {
   if (s == NULL) {
     return NULL;
   }
+  s = mirror_x(s);
 
   ResourceMark rm;
   CodeBuffer cb(s->entry_point(), aarch64_code_length);
@@ -137,6 +138,7 @@ VtableStub* VtableStubs::create_itable_stub(int itable_index) {
   if (s == NULL) {
     return NULL;
   }
+  s = mirror_x(s);
 
   ResourceMark rm;
   CodeBuffer cb(s->entry_point(), code_length);
diff --git a/hotspot/src/os/bsd/vm/os_bsd.cpp b/hotspot/src/os/bsd/vm/os_bsd.cpp
index 522ea6a4..ae373aba 100644
--- a/hotspot/src/os/bsd/vm/os_bsd.cpp
+++ b/hotspot/src/os/bsd/vm/os_bsd.cpp
@@ -156,6 +156,131 @@ static pid_t _initial_pid = 0;
 static int SR_signum = SIGUSR2;
 sigset_t SR_sigset;
 
+#ifdef __APPLE__
+address os::Bsd::debug_jit_mapping_mirrored = NULL;
+
+__attribute__((naked,noinline,optnone))
+static char* BreakGetJITMapping(size_t bytes) {
+    asm("brk #0x69 \n"
+        "ret");
+}
+
+static bool DeviceRequiresTXMWorkaround(bool condition) {
+    if(!condition) {
+        return false;
+    }
+    // 1. Query value from globally-exported DeviceHasTXM symbol from the host app
+    bool (*DeviceHasTXM)(void) = (bool (*)(void))dlsym(RTLD_DEFAULT, "DeviceHasTXM");
+    if(DeviceHasTXM) {
+        return DeviceHasTXM();
+    }
+    printf("[JIT26] DeviceHasTXM() is missing. Looking for value from XNU_HAS_TXM\n");
+    // 2. Query value from an env var
+    const char *value = ::getenv("XNU_HAS_TXM");
+    return value && !strcmp(value, "1");
+}
+
+static char* get_debug_jit_mapping(size_t bytes) {
+    // the map we got has debuggable flag, r-x, setup mirrored map
+    vm_address_t buf_rx = 0;
+    if(MirrorMappedCodeCache) {
+        if(DeviceRequiresTXMWorkaround(__builtin_available(iOS 19.0, *))) {
+            printf("[JIT26] Requesting %zu MB for JIT mapping\n", bytes/ (1024 * 1024));
+            buf_rx = (vm_address_t)BreakGetJITMapping(bytes);
+        }
+        if(buf_rx) {
+            printf("[JIT26] Got JIT mapping %p from debugger\n", (void*)buf_rx);
+        } else {
+            buf_rx = (vm_address_t)mmap(NULL, bytes, PROT_READ | PROT_EXEC, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+        }
+        if(buf_rx == (vm_address_t)MAP_FAILED) {
+            printf("[JIT26] Failed to allocate RX region\n");
+            return NULL;
+        }
+        // on 17.0-18.x we fault in all pages to make TXM happy
+        if(DeviceRequiresTXMWorkaround(!__builtin_available(iOS 19.0, *))) {
+            if (mprotect((void*)buf_rx, bytes, PROT_READ | PROT_WRITE) != 0) {
+                fprintf(stderr, "[JIT26] Failed to make RX region writable: %d\n", errno);
+                return NULL;
+            }
+            bzero((void*)buf_rx, bytes);
+            if (mprotect((void*)buf_rx, bytes, PROT_READ | PROT_EXEC) != 0) {
+                fprintf(stderr, "[JIT26] Failed to restore RX protection: %d\n", errno);
+                return NULL;
+            }
+        }
+    } else {
+        return NULL;
+    }
+    vm_address_t buf_rw = 0;
+
+    vm_prot_t cur_prot, max_prot;
+
+    kern_return_t ret = vm_remap(mach_task_self(), &buf_rw, bytes, 0,
+                                 VM_FLAGS_ANYWHERE, mach_task_self(), buf_rx, false, &cur_prot,
+                                 &max_prot, VM_INHERIT_NONE);
+    if (ret != KERN_SUCCESS) {
+        fprintf(stderr, "[JIT26] Failed to remap RX region %d\n", ret);
+        ::munmap((void*)buf_rx, bytes);
+        return NULL;
+    }
+
+    // Protect region as RW
+    ret = vm_protect(mach_task_self(), buf_rw, bytes, FALSE,
+                     VM_PROT_READ | VM_PROT_WRITE);
+    if (ret != KERN_SUCCESS) {
+        fprintf(stderr, "[JIT26] Failed to set RW protection %d\n", ret);
+        vm_deallocate(mach_task_self(), buf_rw, bytes);
+        ::munmap((void*)buf_rx, bytes);
+        return NULL;
+    }
+
+    printf("[JIT26] mapping at RW=%p, RX=%p\n", (void*)buf_rw, (void*)buf_rx);
+
+    os::Bsd::debug_jit_mapping_mirrored = (address)buf_rw;
+    return (char *)buf_rx;
+    //os::Bsd::debug_jit_mapping_mirrored = (address)buf_rx;
+    //return (char *)buf_rw;
+}
+
+
+address os::Bsd::mirrored_swap_wx(address addr) {
+    if (!MirrorMappedCodeCache || debug_jit_mapping_mirrored == NULL) {
+        return addr;
+    }
+    address low_bound = CodeCache::low_bound();
+    address high_bound = CodeCache::high_bound();
+    size_t size = high_bound - low_bound;
+    if(addr >= low_bound && addr < high_bound) {
+        return addr + (debug_jit_mapping_mirrored - low_bound);
+    } else if (addr >= debug_jit_mapping_mirrored && addr < debug_jit_mapping_mirrored + size) {
+        return addr + (low_bound - debug_jit_mapping_mirrored);
+    }
+    return addr;
+}
+// get_debug_jit_mapping returns rx map
+address os::Bsd::mirrored_find_rw(address addr) {
+    if(!MirrorMappedCodeCache || debug_jit_mapping_mirrored == NULL) return addr;
+    address low_bound = CodeCache::low_bound();
+    address high_bound = CodeCache::high_bound();
+    if(addr >= low_bound && addr < high_bound) {
+        return addr + (debug_jit_mapping_mirrored - low_bound);
+    }
+    return addr;
+}
+address os::Bsd::mirrored_find_rx(address addr) {
+    if(!MirrorMappedCodeCache || debug_jit_mapping_mirrored == NULL) return addr;
+    address low_bound = CodeCache::low_bound();
+    address high_bound = CodeCache::high_bound();
+    size_t size = high_bound - low_bound;
+    if(addr >= low_bound && addr < high_bound) {
+        return addr;
+    } else if(addr >= debug_jit_mapping_mirrored && addr < debug_jit_mapping_mirrored + size) {
+        return addr + (low_bound - debug_jit_mapping_mirrored);
+    }
+    return addr;
+}
+#endif
 
 ////////////////////////////////////////////////////////////////////////////////
 // utility functions
@@ -2223,7 +2348,7 @@ bool os::pd_commit_memory(char* addr, size_t size, bool exec) {
     return true;
   }
 #elif defined(__APPLE__)
-  if (::mprotect(addr, size, prot) == 0) {
+  if ((exec && MirrorMappedCodeCache) || ::mprotect(addr, size, prot) == 0) {
     return true;
   }
 #else
@@ -2392,6 +2517,15 @@ static int anon_munmap(char * addr, size_t size) {
 char* os::pd_reserve_memory(size_t bytes, char* requested_addr,
                          size_t alignment_hint,
                          bool executable) {
+#ifdef __APPLE__
+    if(executable) {
+        assert(!requested_addr, "JIT region should not have fixed address");
+        char *map = get_debug_jit_mapping(bytes);
+        if(map) {
+            return map;
+        }
+    }
+#endif
   return anon_mmap(requested_addr, bytes, (requested_addr != NULL), executable);
 }
 
diff --git a/hotspot/src/os/bsd/vm/os_bsd.hpp b/hotspot/src/os/bsd/vm/os_bsd.hpp
index f67d192d..b2bfbfcc 100644
--- a/hotspot/src/os/bsd/vm/os_bsd.hpp
+++ b/hotspot/src/os/bsd/vm/os_bsd.hpp
@@ -25,6 +25,16 @@
 #ifndef OS_BSD_VM_OS_BSD_HPP
 #define OS_BSD_VM_OS_BSD_HPP
 
+#ifdef __APPLE__
+#define mirror_w(x) ((decltype(x))os::Bsd::mirrored_find_rw((address)x))
+#define mirror_x(x) ((decltype(x))os::Bsd::mirrored_find_rx((address)x))
+#define mirror_w_set(x) *mirror_w(&(x))
+#else
+#define mirror_w(x) (x)
+#define mirror_x(x) (x)
+#define mirror_w_set(x) (x)
+#endif
+
 // Bsd_OS defines the interface to Bsd operating systems
 
 // Information about the protection of the page at address '0' on this os.
@@ -86,6 +96,12 @@ class Bsd {
   static bool hugetlbfs_sanity_check(bool warn, size_t page_size);
 
  public:
+#ifdef __APPLE__
+  static address debug_jit_mapping_mirrored;
+  static address mirrored_swap_wx(address addr);
+  static address mirrored_find_rw(address addr);
+  static address mirrored_find_rx(address addr);
+#endif
 
   static void init_thread_fpu_state();
   static pthread_t main_thread(void)                                { return _main_thread; }
diff --git a/hotspot/src/os_cpu/bsd_aarch64/vm/atomic_bsd_aarch64.inline.hpp b/hotspot/src/os_cpu/bsd_aarch64/vm/atomic_bsd_aarch64.inline.hpp
index 08d7bd81..17427ec6 100644
--- a/hotspot/src/os_cpu/bsd_aarch64/vm/atomic_bsd_aarch64.inline.hpp
+++ b/hotspot/src/os_cpu/bsd_aarch64/vm/atomic_bsd_aarch64.inline.hpp
@@ -35,22 +35,22 @@
 #define READ_MEM_BARRIER  __atomic_thread_fence(__ATOMIC_ACQUIRE);
 #define WRITE_MEM_BARRIER __atomic_thread_fence(__ATOMIC_RELEASE);
 
-inline void Atomic::store    (jbyte    store_value, jbyte*    dest) { *dest = store_value; }
-inline void Atomic::store    (jshort   store_value, jshort*   dest) { *dest = store_value; }
-inline void Atomic::store    (jint     store_value, jint*     dest) { *dest = store_value; }
-inline void Atomic::store_ptr(intptr_t store_value, intptr_t* dest) { *dest = store_value; }
-inline void Atomic::store_ptr(void*    store_value, void*     dest) { *(void**)dest = store_value; }
+inline void Atomic::store    (jbyte    store_value, jbyte*    dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store    (jshort   store_value, jshort*   dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store    (jint     store_value, jint*     dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store_ptr(intptr_t store_value, intptr_t* dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store_ptr(void*    store_value, void*     dest) { *(void**)mirror_w(dest) = store_value; }
 
-inline void Atomic::store    (jbyte    store_value, volatile jbyte*    dest) { *dest = store_value; }
-inline void Atomic::store    (jshort   store_value, volatile jshort*   dest) { *dest = store_value; }
-inline void Atomic::store    (jint     store_value, volatile jint*     dest) { *dest = store_value; }
-inline void Atomic::store_ptr(intptr_t store_value, volatile intptr_t* dest) { *dest = store_value; }
-inline void Atomic::store_ptr(void*    store_value, volatile void*     dest) { *(void* volatile *)dest = store_value; }
+inline void Atomic::store    (jbyte    store_value, volatile jbyte*    dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store    (jshort   store_value, volatile jshort*   dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store    (jint     store_value, volatile jint*     dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store_ptr(intptr_t store_value, volatile intptr_t* dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store_ptr(void*    store_value, volatile void*     dest) { *(void* volatile *)mirror_w(dest) = store_value; }
 
 
 inline jint Atomic::add(jint add_value, volatile jint* dest)
 {
- return __sync_add_and_fetch(dest, add_value);
+ return __sync_add_and_fetch(mirror_w(dest), add_value);
 }
 
 inline void Atomic::inc(volatile jint* dest)
@@ -75,7 +75,7 @@ inline void Atomic::dec_ptr(volatile void* dest)
 
 inline jint Atomic::xchg (jint exchange_value, volatile jint* dest)
 {
-  jint res = __sync_lock_test_and_set (dest, exchange_value);
+  jint res = __sync_lock_test_and_set (mirror_w(dest), exchange_value);
   FULL_MEM_BARRIER;
   return res;
 }
@@ -88,15 +88,15 @@ inline void* Atomic::xchg_ptr(void* exchange_value, volatile void* dest)
 
 inline jint Atomic::cmpxchg (jint exchange_value, volatile jint* dest, jint compare_value)
 {
- return __sync_val_compare_and_swap(dest, compare_value, exchange_value);
+ return __sync_val_compare_and_swap(mirror_w(dest), compare_value, exchange_value);
 }
 
-inline void Atomic::store (jlong store_value, jlong* dest) { *dest = store_value; }
-inline void Atomic::store (jlong store_value, volatile jlong* dest) { *dest = store_value; }
+inline void Atomic::store (jlong store_value, jlong* dest) { *mirror_w(dest) = store_value; }
+inline void Atomic::store (jlong store_value, volatile jlong* dest) { *mirror_w(dest) = store_value; }
 
 inline intptr_t Atomic::add_ptr(intptr_t add_value, volatile intptr_t* dest)
 {
- return __sync_add_and_fetch(dest, add_value);
+ return __sync_add_and_fetch(mirror_w(dest), add_value);
 }
 
 inline void* Atomic::add_ptr(intptr_t add_value, volatile void* dest)
@@ -116,19 +116,19 @@ inline void Atomic::dec_ptr(volatile intptr_t* dest)
 
 inline intptr_t Atomic::xchg_ptr(intptr_t exchange_value, volatile intptr_t* dest)
 {
-  intptr_t res = __sync_lock_test_and_set (dest, exchange_value);
+  intptr_t res = __sync_lock_test_and_set (mirror_w(dest), exchange_value);
   FULL_MEM_BARRIER;
   return res;
 }
 
 inline jlong Atomic::cmpxchg (jlong exchange_value, volatile jlong* dest, jlong compare_value)
 {
- return __sync_val_compare_and_swap(dest, compare_value, exchange_value);
+ return __sync_val_compare_and_swap(mirror_w(dest), compare_value, exchange_value);
 }
 
 inline intptr_t Atomic::cmpxchg_ptr(intptr_t exchange_value, volatile intptr_t* dest, intptr_t compare_value)
 {
- return __sync_val_compare_and_swap(dest, compare_value, exchange_value);
+ return __sync_val_compare_and_swap(mirror_w(dest), compare_value, exchange_value);
 }
 
 inline void* Atomic::cmpxchg_ptr(void* exchange_value, volatile void* dest, void* compare_value)
diff --git a/hotspot/src/os_cpu/bsd_aarch64/vm/orderAccess_bsd_aarch64.inline.hpp b/hotspot/src/os_cpu/bsd_aarch64/vm/orderAccess_bsd_aarch64.inline.hpp
index 0c5a9839..8a01b381 100644
--- a/hotspot/src/os_cpu/bsd_aarch64/vm/orderAccess_bsd_aarch64.inline.hpp
+++ b/hotspot/src/os_cpu/bsd_aarch64/vm/orderAccess_bsd_aarch64.inline.hpp
@@ -78,54 +78,54 @@ inline void*    OrderAccess::load_ptr_acquire(const volatile void* p)
 { void* data; __atomic_load((void* const volatile *)p, &data, __ATOMIC_ACQUIRE); return data; }
 
 inline void     OrderAccess::release_store(volatile jbyte*   p, jbyte   v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile jshort*  p, jshort  v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile jint*    p, jint    v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile jlong*   p, jlong   v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile jubyte*  p, jubyte  v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile jushort* p, jushort v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile juint*   p, juint   v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile julong*  p, julong  v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile jfloat*  p, jfloat  v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store(volatile jdouble* p, jdouble v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store_ptr(volatile intptr_t* p, intptr_t v)
-{ __atomic_store(p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELEASE); }
 inline void     OrderAccess::release_store_ptr(volatile void*     p, void*    v)
-{ __atomic_store((void* volatile *)p, &v, __ATOMIC_RELEASE); }
+{ __atomic_store((void* volatile *)mirror_w(p), &v, __ATOMIC_RELEASE); }
 
 inline void     OrderAccess::store_fence(jbyte*   p, jbyte   v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(jshort*  p, jshort  v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(jint*    p, jint    v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(jlong*   p, jlong   v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(jubyte*  p, jubyte  v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(jushort* p, jushort v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(juint*   p, juint   v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(julong*  p, julong  v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(jfloat*  p, jfloat  v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_fence(jdouble* p, jdouble v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_ptr_fence(intptr_t* p, intptr_t v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 inline void     OrderAccess::store_ptr_fence(void**    p, void*    v)
-{ __atomic_store(p, &v, __ATOMIC_RELAXED); fence(); }
+{ __atomic_store(mirror_w(p), &v, __ATOMIC_RELAXED); fence(); }
 
 inline void     OrderAccess::release_store_fence(volatile jbyte*   p, jbyte   v) { release_store(p, v); fence(); }
 inline void     OrderAccess::release_store_fence(volatile jshort*  p, jshort  v) { release_store(p, v); fence(); }
diff --git a/hotspot/src/os_cpu/bsd_aarch64/vm/os_bsd_aarch64.cpp b/hotspot/src/os_cpu/bsd_aarch64/vm/os_bsd_aarch64.cpp
index 7cdd2fa5..a1d34fb2 100644
--- a/hotspot/src/os_cpu/bsd_aarch64/vm/os_bsd_aarch64.cpp
+++ b/hotspot/src/os_cpu/bsd_aarch64/vm/os_bsd_aarch64.cpp
@@ -211,7 +211,7 @@ JVM_handle_bsd_signal(int sig,
                         int abort_if_unrecognized) {
   ucontext_t* uc = (ucontext_t*) ucVoid;
   //Hopefully placing W^X handing here is safe enough, maybe check repeat?
-  if (sig == SIGBUS) {
+  if (!MirrorMappedCodeCache && sig == SIGBUS) {
     address pc = (address) os::Bsd::ucontext_get_pc(uc);
     // build machine, do not run because of panic on M4
     if (pc == info->si_addr) {
diff --git a/hotspot/src/share/vm/asm/assembler.cpp b/hotspot/src/share/vm/asm/assembler.cpp
index 0eadb5dc..f52de65b 100644
--- a/hotspot/src/share/vm/asm/assembler.cpp
+++ b/hotspot/src/share/vm/asm/assembler.cpp
@@ -56,7 +56,7 @@ void AbstractAssembler::set_code_section(CodeSection* cs) {
   assert(cs->outer() == code_section()->outer(), "sanity");
   assert(cs->is_allocated(), "need to pre-allocate this section");
   cs->clear_mark();  // new assembly into this section kills old mark
-  _code_section = cs;
+  mirror_w_set(_code_section) = cs;
 }
 
 // Inform CodeBuffer that incoming code and relocation will be for stubs
@@ -91,7 +91,7 @@ address AbstractAssembler::start_a_const(int required_space, int required_align)
     end = cs->end();  // refresh pointer
   }
   if (pad > 0) {
-    while (--pad >= 0) { *end++ = 0; }
+    while (--pad >= 0) { *mirror_w(end++) = 0; }
     cs->set_end(end);
   }
   set_code_section(cs);
diff --git a/hotspot/src/share/vm/asm/codeBuffer.cpp b/hotspot/src/share/vm/asm/codeBuffer.cpp
index d94ac406..e9ef201d 100644
--- a/hotspot/src/share/vm/asm/codeBuffer.cpp
+++ b/hotspot/src/share/vm/asm/codeBuffer.cpp
@@ -358,7 +358,7 @@ void CodeSection::relocate(address at, RelocationHolder const& spec, int format)
   }
 
   // If it's a simple reloc with no data, we'll just write (rtype | offset).
-  (*end) = relocInfo(rtype, offset, format);
+  (*mirror_w(end)) = relocInfo(rtype, offset, format);
 
   // If it has data, insert the prefix, as (data_prefix_tag | data1), data2.
   end->initialize(this, reloc);
@@ -644,7 +644,7 @@ csize_t CodeBuffer::copy_relocations_to(CodeBlob* dest) const {
         }
         if (buf != NULL) {
           assert(buf_offset + (csize_t)sizeof(filler) <= buf_limit, "filler in bounds");
-          *(relocInfo*)(buf+buf_offset) = filler;
+          *(relocInfo*)mirror_w(buf+buf_offset) = filler;
         }
         buf_offset += sizeof(filler);
       }
@@ -663,10 +663,10 @@ csize_t CodeBuffer::copy_relocations_to(CodeBlob* dest) const {
       if (buf_offset % HeapWordSize == 0) {
         // Use wordwise copies if possible:
         Copy::disjoint_words((HeapWord*)lstart,
-                             (HeapWord*)(buf+buf_offset),
+                             (HeapWord*)mirror_w(buf+buf_offset),
                              (lsize + HeapWordSize-1) / HeapWordSize);
       } else {
-        Copy::conjoint_jbytes(lstart, buf+buf_offset, lsize);
+        Copy::conjoint_jbytes(lstart, mirror_w(buf+buf_offset), lsize);
       }
     }
     buf_offset += lsize;
@@ -677,7 +677,7 @@ csize_t CodeBuffer::copy_relocations_to(CodeBlob* dest) const {
     if (buf != NULL) {
       relocInfo padding = relocInfo(relocInfo::none, 0);
       assert(buf_offset + (csize_t)sizeof(padding) <= buf_limit, "padding in bounds");
-      *(relocInfo*)(buf+buf_offset) = padding;
+      *(relocInfo*)mirror_w(buf+buf_offset) = padding;
     }
     buf_offset += sizeof(relocInfo);
   }
@@ -737,13 +737,13 @@ void CodeBuffer::relocate_code_to(CodeBuffer* dest) const {
     // Copy the code as aligned machine words.
     // This may also include an uninitialized partial word at the end.
     Copy::disjoint_words((HeapWord*)cs->start(),
-                         (HeapWord*)dest_cs->start(),
+                         (HeapWord*)mirror_w(dest_cs->start()),
                          wsize / HeapWordSize);
 
     if (dest->blob() == NULL) {
       // Destination is a final resting place, not just another buffer.
       // Normalize uninitialized bytes in the final padding.
-      Copy::fill_to_bytes(dest_cs->end(), dest_cs->remaining(),
+      Copy::fill_to_bytes(mirror_w(dest_cs->end()), dest_cs->remaining(),
                           Assembler::code_fill_byte());
     }
     // Keep track of the highest filled address
@@ -777,7 +777,7 @@ void CodeBuffer::relocate_code_to(CodeBuffer* dest) const {
   if (dest->blob() == NULL && dest_filled != NULL) {
     // Destination is a final resting place, not just another buffer.
     // Normalize uninitialized bytes in the final padding.
-    Copy::fill_to_bytes(dest_filled, dest_end - dest_filled,
+    Copy::fill_to_bytes(mirror_w(dest_filled), dest_end - dest_filled,
                         Assembler::code_fill_byte());
 
   }
@@ -906,7 +906,7 @@ void CodeBuffer::expand(CodeSection* which_cs, csize_t amount) {
   cb.set_blob(NULL);
 
   // Zap the old code buffer contents, to avoid mistakenly using them.
-  debug_only(Copy::fill_to_bytes(bxp->_total_start, bxp->_total_size,
+  debug_only(Copy::fill_to_bytes(mirror_w(bxp->_total_start), bxp->_total_size,
                                  badCodeHeapFreeVal));
 
   _decode_begin = NULL;  // sanity
diff --git a/hotspot/src/share/vm/asm/codeBuffer.hpp b/hotspot/src/share/vm/asm/codeBuffer.hpp
index 02b619ad..3fede168 100644
--- a/hotspot/src/share/vm/asm/codeBuffer.hpp
+++ b/hotspot/src/share/vm/asm/codeBuffer.hpp
@@ -27,6 +27,7 @@
 
 #include "code/oopRecorder.hpp"
 #include "code/relocInfo.hpp"
+#include "runtime/os.hpp"
 #include "utilities/debug.hpp"
 
 class CodeStrings;
@@ -118,6 +119,7 @@ class CodeSection VALUE_OBJ_CLASS_SPEC {
   }
 
   void initialize(address start, csize_t size = 0) {
+    start = mirror_x(start);
     assert(_start == NULL, "only one init step, please");
     _start         = start;
     _mark          = NULL;
@@ -193,14 +195,14 @@ class CodeSection VALUE_OBJ_CLASS_SPEC {
   }
 
   // Code emission
-  void emit_int8 ( int8_t  x)  { *((int8_t*)  end()) = x; set_end(end() + sizeof(int8_t)); }
-  void emit_int16( int16_t x)  { *((int16_t*) end()) = x; set_end(end() + sizeof(int16_t)); }
-  void emit_int32( int32_t x)  { *((int32_t*) end()) = x; set_end(end() + sizeof(int32_t)); }
-  void emit_int64( int64_t x)  { *((int64_t*) end()) = x; set_end(end() + sizeof(int64_t)); }
+  void emit_int8 ( int8_t  x)  { *((int8_t*)  mirror_w(end())) = x; set_end(end() + sizeof(int8_t)); }
+  void emit_int16( int16_t x)  { *((int16_t*) mirror_w(end())) = x; set_end(end() + sizeof(int16_t)); }
+  void emit_int32( int32_t x)  { *((int32_t*) mirror_w(end())) = x; set_end(end() + sizeof(int32_t)); }
+  void emit_int64( int64_t x)  { *((int64_t*) mirror_w(end())) = x; set_end(end() + sizeof(int64_t)); }
 
-  void emit_float( jfloat  x)  { *((jfloat*)  end()) = x; set_end(end() + sizeof(jfloat)); }
-  void emit_double(jdouble x)  { *((jdouble*) end()) = x; set_end(end() + sizeof(jdouble)); }
-  void emit_address(address x) { *((address*) end()) = x; set_end(end() + sizeof(address)); }
+  void emit_float( jfloat  x)  { *((jfloat*)  mirror_w(end())) = x; set_end(end() + sizeof(jfloat)); }
+  void emit_double(jdouble x)  { *((jdouble*) mirror_w(end())) = x; set_end(end() + sizeof(jdouble)); }
+  void emit_address(address x) { *((address*) mirror_w(end())) = x; set_end(end() + sizeof(address)); }
 
   // Share a scratch buffer for relocinfo.  (Hacky; saves a resource allocation.)
   void initialize_shared_locs(relocInfo* buf, int length);
@@ -383,6 +385,7 @@ class CodeBuffer: public StackObj {
   }
 
   void initialize(address code_start, csize_t code_size) {
+    code_start = mirror_x(code_start);
     _consts.initialize_outer(this,  SECT_CONSTS);
     _insts.initialize_outer(this,   SECT_INSTS);
     _stubs.initialize_outer(this,   SECT_STUBS);
diff --git a/hotspot/src/share/vm/code/codeBlob.cpp b/hotspot/src/share/vm/code/codeBlob.cpp
index aff2aaf0..930b9956 100644
--- a/hotspot/src/share/vm/code/codeBlob.cpp
+++ b/hotspot/src/share/vm/code/codeBlob.cpp
@@ -148,10 +148,10 @@ void CodeBlob::set_oop_maps(OopMapSet* p) {
   // chunk of memory, its your job to free it.
   if (p != NULL) {
     // We need to allocate a chunk big enough to hold the OopMapSet and all of its OopMaps
-    _oop_maps = (OopMapSet* )NEW_C_HEAP_ARRAY(unsigned char, p->heap_size(), mtCode);
+    mirror_w_set(_oop_maps) = (OopMapSet* )NEW_C_HEAP_ARRAY(unsigned char, p->heap_size(), mtCode);
     p->copy_to((address)_oop_maps);
   } else {
-    _oop_maps = NULL;
+    mirror_w_set(_oop_maps) = NULL;
   }
 }
 
@@ -187,9 +187,9 @@ void CodeBlob::trace_new_stub(CodeBlob* stub, const char* name1, const char* nam
 void CodeBlob::flush() {
   if (_oop_maps) {
     FREE_C_HEAP_ARRAY(unsigned char, _oop_maps, mtCode);
-    _oop_maps = NULL;
+    mirror_w_set(_oop_maps) = NULL;
   }
-  _strings.free();
+  mirror_w(this)->_strings.free();
 }
 
 
@@ -228,7 +228,7 @@ BufferBlob* BufferBlob::create(const char* name, int buffer_size) {
   // Track memory usage statistic after releasing CodeCache_lock
   MemoryService::track_code_cache_memory_usage();
 
-  return blob;
+  return mirror_x(blob);
 }
 
 
@@ -249,13 +249,13 @@ BufferBlob* BufferBlob::create(const char* name, CodeBuffer* cb) {
   // Track memory usage statistic after releasing CodeCache_lock
   MemoryService::track_code_cache_memory_usage();
 
-  return blob;
+  return mirror_x(blob);
 }
 
 
 void* BufferBlob::operator new(size_t s, unsigned size, bool is_critical) throw() {
   void* p = CodeCache::allocate(size, is_critical);
-  return p;
+  return mirror_w(p);
 }
 
 
@@ -276,7 +276,7 @@ void BufferBlob::free( BufferBlob *blob ) {
 
 AdapterBlob::AdapterBlob(int size, CodeBuffer* cb) :
   BufferBlob("I2C/C2I adapters", size, cb) {
-  CodeCache::commit(this);
+  CodeCache::commit(mirror_x(this));
 }
 
 AdapterBlob* AdapterBlob::create(CodeBuffer* cb) {
@@ -289,12 +289,13 @@ AdapterBlob* AdapterBlob::create(CodeBuffer* cb) {
     // The parameter 'true' indicates a critical memory allocation.
     // This means that CodeCacheMinimumFreeSpace is used, if necessary
     const bool is_critical = true;
+    cb = mirror_w(cb);
     blob = new (size, is_critical) AdapterBlob(size, cb);
   }
   // Track memory usage statistic after releasing CodeCache_lock
   MemoryService::track_code_cache_memory_usage();
 
-  return blob;
+  return mirror_x(blob);
 }
 
 VtableBlob::VtableBlob(const char* name, int size) :
@@ -317,7 +318,7 @@ VtableBlob* VtableBlob::create(const char* name, int buffer_size) {
   // Track memory usage statistic after releasing CodeCache_lock
   MemoryService::track_code_cache_memory_usage();
 
-  return blob;
+  return mirror_x(blob);
 }
 
 //----------------------------------------------------------------------------------------------------
@@ -341,7 +342,7 @@ MethodHandlesAdapterBlob* MethodHandlesAdapterBlob::create(int buffer_size) {
   // Track memory usage statistic after releasing CodeCache_lock
   MemoryService::track_code_cache_memory_usage();
 
-  return blob;
+  return mirror_x(blob);
 }
 
 
@@ -380,21 +381,21 @@ RuntimeStub* RuntimeStub::new_runtime_stub(const char* stub_name,
 
   trace_new_stub(stub, "RuntimeStub - ", stub_name);
 
-  return stub;
+  return mirror_x(stub);
 }
 
 
 void* RuntimeStub::operator new(size_t s, unsigned size) throw() {
   void* p = CodeCache::allocate(size, true);
   if (!p) fatal("Initial size of CodeCache is too small");
-  return p;
+  return mirror_w(p);
 }
 
 // operator new shared by all singletons:
 void* SingletonBlob::operator new(size_t s, unsigned size) throw() {
   void* p = CodeCache::allocate(size, true);
   if (!p) fatal("Initial size of CodeCache is too small");
-  return p;
+  return mirror_w(p);
 }
 
 
@@ -445,7 +446,7 @@ DeoptimizationBlob* DeoptimizationBlob::create(
 
   trace_new_stub(blob, "DeoptimizationBlob");
 
-  return blob;
+  return mirror_x(blob);
 }
 
 
@@ -478,7 +479,7 @@ UncommonTrapBlob* UncommonTrapBlob::create(
 
   trace_new_stub(blob, "UncommonTrapBlob");
 
-  return blob;
+  return mirror_x(blob);
 }
 
 
@@ -514,7 +515,7 @@ ExceptionBlob* ExceptionBlob::create(
 
   trace_new_stub(blob, "ExceptionBlob");
 
-  return blob;
+  return mirror_x(blob);
 }
 
 
@@ -549,7 +550,7 @@ SafepointBlob* SafepointBlob::create(
 
   trace_new_stub(blob, "SafepointBlob");
 
-  return blob;
+  return mirror_x(blob);
 }
 
 
diff --git a/hotspot/src/share/vm/code/codeBlob.hpp b/hotspot/src/share/vm/code/codeBlob.hpp
index f956f421..f11994f1 100644
--- a/hotspot/src/share/vm/code/codeBlob.hpp
+++ b/hotspot/src/share/vm/code/codeBlob.hpp
@@ -121,8 +121,8 @@ class CodeBlob VALUE_OBJ_CLASS_SPEC {
   nmethod* as_nmethod_or_null()                  { return is_nmethod() ? (nmethod*) this : NULL; }
 
   // Boundaries
-  address    header_begin() const                { return (address)    this; }
-  address    header_end() const                  { return ((address)   this) + _header_size; };
+  address    header_begin() const                { return (address)    mirror_x(this); }
+  address    header_end() const                  { return ((address)   mirror_x(this)) + _header_size; };
   relocInfo* relocation_begin() const            { return (relocInfo*) header_end(); };
   relocInfo* relocation_end() const              { return (relocInfo*)(header_end()   + _relocation_size); }
   address    content_begin() const               { return (address)    header_begin() + _content_offset; }
@@ -176,14 +176,14 @@ class CodeBlob VALUE_OBJ_CLASS_SPEC {
 
   // Frame support
   int  frame_size() const                        { return _frame_size; }
-  void set_frame_size(int size)                  { _frame_size = size; }
+  void set_frame_size(int size)                  { mirror_w_set(_frame_size) = size; }
 
   // Returns true, if the next frame is responsible for GC'ing oops passed as arguments
   virtual bool caller_must_gc_arguments(JavaThread* thread) const { return false; }
 
   // Naming
   const char* name() const                       { return _name; }
-  void set_name(const char* name)                { _name = name; }
+  void set_name(const char* name)                { mirror_w_set(_name) = name; }
 
   // Debugging
   virtual void verify();
@@ -203,7 +203,7 @@ class CodeBlob VALUE_OBJ_CLASS_SPEC {
 
   // Transfer ownership of comments to this CodeBlob
   void set_strings(CodeStrings& strings) {
-    _strings.assign(strings);
+    mirror_w(this)->_strings.assign(strings);
   }
 };
 
@@ -430,7 +430,7 @@ class DeoptimizationBlob: public SingletonBlob {
   // model exception paths in a way that keeps these registers free so
   // there may be live values in those registers during deopt.
   void set_unpack_with_exception_in_tls_offset(int offset) {
-    _unpack_with_exception_in_tls = offset;
+    mirror_w_set(_unpack_with_exception_in_tls) = offset;
     assert(code_contains(code_begin() + _unpack_with_exception_in_tls), "must be PC inside codeblob");
   }
   address unpack_with_exception_in_tls() const   { return code_begin() + _unpack_with_exception_in_tls; }
diff --git a/hotspot/src/share/vm/code/dependencies.cpp b/hotspot/src/share/vm/code/dependencies.cpp
index d1fe08b5..27b50c8d 100644
--- a/hotspot/src/share/vm/code/dependencies.cpp
+++ b/hotspot/src/share/vm/code/dependencies.cpp
@@ -245,7 +245,7 @@ void Dependencies::copy_to(nmethod* nm) {
   address end = nm->dependencies_end();
   guarantee(end - beg >= (ptrdiff_t) size_in_bytes(), "bad sizing");
   Copy::disjoint_words((HeapWord*) content_bytes(),
-                       (HeapWord*) beg,
+                       (HeapWord*) mirror_w(beg),
                        size_in_bytes() / sizeof(HeapWord));
   assert(size_in_bytes() % sizeof(HeapWord) == 0, "copy by words");
 }
diff --git a/hotspot/src/share/vm/code/exceptionHandlerTable.cpp b/hotspot/src/share/vm/code/exceptionHandlerTable.cpp
index 224e164d..77919de2 100644
--- a/hotspot/src/share/vm/code/exceptionHandlerTable.cpp
+++ b/hotspot/src/share/vm/code/exceptionHandlerTable.cpp
@@ -102,7 +102,7 @@ void ExceptionHandlerTable::add_subtable(
 
 void ExceptionHandlerTable::copy_to(nmethod* nm) {
   assert(size_in_bytes() == nm->handler_table_size(), "size of space allocated in nmethod incorrect");
-  memmove(nm->handler_table_begin(), _table, size_in_bytes());
+  memmove(mirror_w(nm->handler_table_begin()), _table, size_in_bytes());
 }
 
 
@@ -208,7 +208,7 @@ ImplicitExceptionTable::ImplicitExceptionTable(const nmethod* nm) {
 void ImplicitExceptionTable::copy_to( nmethod* nm ) {
   assert(size_in_bytes() <= nm->nul_chk_table_size(), "size of space allocated in nmethod incorrect");
   if (len() != 0) {
-    implicit_null_entry* nmdata = (implicit_null_entry*)nm->nul_chk_table_begin();
+    implicit_null_entry* nmdata = (implicit_null_entry*)mirror_w(nm->nul_chk_table_begin());
     // store the length in the first uint
     nmdata[0] = _len;
     nmdata++;
diff --git a/hotspot/src/share/vm/code/icBuffer.cpp b/hotspot/src/share/vm/code/icBuffer.cpp
index 0fe6e0d1..8b117444 100644
--- a/hotspot/src/share/vm/code/icBuffer.cpp
+++ b/hotspot/src/share/vm/code/icBuffer.cpp
@@ -74,7 +74,7 @@ void ICStub::set_stub(CompiledIC *ic, void* cached_val, address dest_addr) {
   // We cannot store a pointer to the 'ic' object, since it is resource allocated. Instead we
   // store the location of the inline cache. Then we have enough information recreate the CompiledIC
   // object when we need to remove the stub.
-  _ic_site = ic->instruction_address();
+  mirror_w_set(_ic_site) = ic->instruction_address();
 
   // Assemble new stub
   InlineCacheBuffer::assemble_ic_buffer_code(code_begin(), cached_val, dest_addr);
@@ -87,7 +87,7 @@ void ICStub::clear() {
   if (CompiledIC::is_icholder_entry(destination())) {
     InlineCacheBuffer::queue_for_release((CompiledICHolder*)cached_value());
   }
-  _ic_site = NULL;
+  mirror_w_set(_ic_site) = NULL;
 }
 
 
diff --git a/hotspot/src/share/vm/code/icBuffer.hpp b/hotspot/src/share/vm/code/icBuffer.hpp
index 471706aa..65835656 100644
--- a/hotspot/src/share/vm/code/icBuffer.hpp
+++ b/hotspot/src/share/vm/code/icBuffer.hpp
@@ -50,7 +50,7 @@ class ICStub: public Stub {
   friend class ICStubInterface;
   // This will be called only by ICStubInterface
   void    initialize(int size,
-                     CodeStrings strings)        { _size = size; _ic_site = NULL; }
+                     CodeStrings strings)        { mirror_w_set(_size) = size; mirror_w_set(_ic_site) = NULL; }
   void    finalize(); // called when a method is removed
 
   // General info
diff --git a/hotspot/src/share/vm/code/nmethod.cpp b/hotspot/src/share/vm/code/nmethod.cpp
index 55a743ae..b94572c7 100644
--- a/hotspot/src/share/vm/code/nmethod.cpp
+++ b/hotspot/src/share/vm/code/nmethod.cpp
@@ -314,7 +314,7 @@ void PcDescCache::reset_to(PcDesc* initial_pc_desc) {
   // reset the cache by filling it with benign (non-null) values
   assert(initial_pc_desc->pc_offset() < 0, "must be sentinel");
   for (int i = 0; i < cache_size; i++)
-    _pc_descs[i] = initial_pc_desc;
+    _pc_descs[i] = mirror_x(initial_pc_desc);
 }
 
 PcDesc* PcDescCache::find_pc_desc(int pc_offset, bool approximate) {
@@ -358,7 +358,7 @@ void PcDescCache::add_pc_desc(PcDesc* pc_desc) {
   // Update the LRU cache by shifting pc_desc forward.
   for (int i = 0; i < cache_size; i++)  {
     PcDesc* next = _pc_descs[i];
-    _pc_descs[i] = pc_desc;
+    mirror_w_set(_pc_descs[i]) = mirror_x(pc_desc);
     pc_desc = next;
   }
 }
@@ -529,6 +529,7 @@ nmethod* nmethod::new_native_nmethod(methodHandle method,
                                             code_buffer, frame_size,
                                             basic_lock_owner_sp_offset,
                                             basic_lock_sp_offset, oop_maps);
+    nm = mirror_x(nm);
     NOT_PRODUCT(if (nm != NULL)  nmethod_stats.note_native_nmethod(nm));
     if (PrintAssembly && nm != NULL) {
       Disassembler::decode(nm);
@@ -563,6 +564,7 @@ nmethod* nmethod::new_dtrace_nmethod(methodHandle method,
 
     nm = new (nmethod_size) nmethod(method(), nmethod_size,
                                     &offsets, code_buffer, frame_size);
+    nm = mirror_x(nm);
 
     NOT_PRODUCT(if (nm != NULL)  nmethod_stats.note_nmethod(nm));
     if (PrintAssembly && nm != NULL) {
@@ -617,6 +619,7 @@ nmethod* nmethod::new_nmethod(methodHandle method,
             nul_chk_table,
             compiler,
             comp_level);
+    nm = mirror_x(nm);
 
     if (nm != NULL) {
       // To make dependency checking during class loading fast, record
@@ -701,15 +704,15 @@ nmethod::nmethod(
     _pc_desc_cache.reset_to(NULL);
     _hotness_counter         = NMethodSweeper::hotness_counter_reset_val();
 
-    code_buffer->copy_values_to(this);
+    code_buffer->copy_values_to(mirror_x(this));
     if (ScavengeRootsInCode) {
       if (detect_scavenge_root_oops()) {
-        CodeCache::add_scavenge_root_nmethod(this);
+        CodeCache::add_scavenge_root_nmethod(mirror_x(this));
       }
-      Universe::heap()->register_nmethod(this);
+      Universe::heap()->register_nmethod(mirror_x(this));
     }
     debug_only(verify_scavenge_root_oops());
-    CodeCache::commit(this);
+    CodeCache::commit(mirror_x(this));
   }
 
   if (PrintNativeNMethods || PrintDebugInfo || PrintRelocations || PrintDependencies) {
@@ -788,15 +791,15 @@ nmethod::nmethod(
     _pc_desc_cache.reset_to(NULL);
     _hotness_counter         = NMethodSweeper::hotness_counter_reset_val();
 
-    code_buffer->copy_values_to(this);
+    code_buffer->copy_values_to(mirror_x(this));
     if (ScavengeRootsInCode) {
       if (detect_scavenge_root_oops()) {
-        CodeCache::add_scavenge_root_nmethod(this);
+        CodeCache::add_scavenge_root_nmethod(mirror_x(this));
       }
-      Universe::heap()->register_nmethod(this);
+      Universe::heap()->register_nmethod(mirror_x(this));
     }
     DEBUG_ONLY(verify_scavenge_root_oops();)
-    CodeCache::commit(this);
+    CodeCache::commit(mirror_x(this));
   }
 
   if (PrintNMethods || PrintDebugInfo || PrintRelocations || PrintDependencies) {
@@ -828,7 +831,7 @@ nmethod::nmethod(
 
 void* nmethod::operator new(size_t size, int nmethod_size) throw() {
   // Not critical, may return null if there is too little continuous memory
-  return CodeCache::allocate(nmethod_size);
+  return mirror_w(CodeCache::allocate(nmethod_size));
 }
 
 nmethod::nmethod(
@@ -904,22 +907,22 @@ nmethod::nmethod(
     _pc_desc_cache.reset_to(scopes_pcs_begin());
 
     // Copy contents of ScopeDescRecorder to nmethod
-    code_buffer->copy_values_to(this);
-    debug_info->copy_to(this);
-    dependencies->copy_to(this);
+    code_buffer->copy_values_to(mirror_x(this));
+    debug_info->copy_to(mirror_x(this));
+    dependencies->copy_to(mirror_x(this));
     if (ScavengeRootsInCode) {
       if (detect_scavenge_root_oops()) {
-        CodeCache::add_scavenge_root_nmethod(this);
+        CodeCache::add_scavenge_root_nmethod(mirror_x(this));
       }
-      Universe::heap()->register_nmethod(this);
+      Universe::heap()->register_nmethod(mirror_x(this));
     }
     debug_only(verify_scavenge_root_oops());
 
-    CodeCache::commit(this);
+    CodeCache::commit(mirror_x(this));
 
     // Copy contents of ExceptionHandlerTable to nmethod
-    handler_table->copy_to(this);
-    nul_chk_table->copy_to(this);
+    handler_table->copy_to(mirror_x(this));
+    nul_chk_table->copy_to(mirror_x(this));
 
     // we use the information of entry points to find out if a method is
     // static or non static
@@ -1042,9 +1045,9 @@ inline void nmethod::initialize_immediate_oop(oop* dest, jobject handle) {
   if (handle == NULL ||
       // As a special case, IC oops are initialized to 1 or -1.
       handle == (jobject) Universe::non_oop_word()) {
-    (*dest) = (oop) handle;
+    (*mirror_w(dest)) = (oop) handle;
   } else {
-    (*dest) = JNIHandles::resolve_non_null(handle);
+    (*mirror_w(dest)) = JNIHandles::resolve_non_null(handle);
   }
 }
 
@@ -1071,7 +1074,7 @@ void nmethod::copy_values(GrowableArray<Metadata*>* array) {
   assert((address)(metadata_begin() + length) <= (address)metadata_end(), "big enough");
   Metadata** dest = metadata_begin();
   for (int index = 0 ; index < length; index++) {
-    dest[index] = array->at(index);
+    mirror_w_set(dest[index]) = array->at(index);
   }
 }
 
@@ -1326,7 +1329,7 @@ void nmethod::increase_unloading_clock() {
 }
 
 void nmethod::set_unloading_clock(unsigned char unloading_clock) {
-  OrderAccess::release_store((volatile jubyte*)&_unloading_clock, unloading_clock);
+  OrderAccess::release_store(mirror_w((volatile jubyte*)&_unloading_clock), unloading_clock);
 }
 
 unsigned char nmethod::unloading_clock() {
@@ -1368,7 +1371,7 @@ void nmethod::make_unloaded(BoolObjectClosure* is_alive, oop cause) {
     if (_method->code() == this) {
       _method->clear_code(); // Break a cycle
     }
-    _method = NULL;            // Clear the method of this dead nmethod
+    mirror_w_set(_method) = NULL;            // Clear the method of this dead nmethod
   }
   // Make the class unloaded - i.e., change state and notify sweeper
   assert(SafepointSynchronize::is_at_safepoint(), "must be at safepoint");
@@ -1382,7 +1385,7 @@ void nmethod::make_unloaded(BoolObjectClosure* is_alive, oop cause) {
   // Unregister must be done before the state change
   Universe::heap()->unregister_nmethod(this);
 
-  _state = unloaded;
+  mirror_w_set(_state) = unloaded;
 
   // Log the unloading.
   log_state_change();
@@ -1400,7 +1403,7 @@ void nmethod::invalidate_osr_method() {
   if (method() != NULL)
     method()->method_holder()->remove_osr_nmethod(this);
   // Set entry as invalid
-  _entry_bci = InvalidOSREntryBci;
+  mirror_w_set(_entry_bci) = InvalidOSREntryBci;
 }
 
 void nmethod::log_state_change() const {
@@ -1493,7 +1496,7 @@ bool nmethod::make_not_entrant_or_zombie(unsigned int state) {
     }
 
     // Change state
-    _state = state;
+    mirror_w_set(_state) = state;
 
     // Log the transition once
     log_state_change();
@@ -1538,7 +1541,7 @@ bool nmethod::make_not_entrant_or_zombie(unsigned int state) {
 #ifdef ASSERT
     // It's no longer safe to access the oops section since zombie
     // nmethods aren't scanned for GC.
-    _oops_are_stale = true;
+    mirror_w_set(_oops_are_stale) = true;
 #endif
      // the Method may be reclaimed by class unloading now that the
      // nmethod is in zombie state
@@ -1697,7 +1700,7 @@ jmethodID nmethod::get_and_cache_jmethod_id() {
   if (_jmethod_id == NULL) {
     // Cache the jmethod_id since it can no longer be looked up once the
     // method itself has been marked for unloading.
-    _jmethod_id = method()->jmethod_id();
+    mirror_w_set(_jmethod_id) = method()->jmethod_id();
   }
   return _jmethod_id;
 }
@@ -2266,7 +2269,7 @@ bool nmethod::test_set_oops_do_mark() {
       nmethod* observed_mark_nmethods = _oops_do_mark_nmethods;
       for (;;) {
         nmethod* required_mark_nmethods = observed_mark_nmethods;
-        _oops_do_mark_link = required_mark_nmethods;
+        mirror_w_set(_oops_do_mark_link) = required_mark_nmethods;
         observed_mark_nmethods = (nmethod*)
           Atomic::cmpxchg_ptr(this, &_oops_do_mark_nmethods, required_mark_nmethods);
         if (observed_mark_nmethods == required_mark_nmethods)
@@ -2296,7 +2299,7 @@ void nmethod::oops_do_marking_epilogue() {
   while (cur != NMETHOD_SENTINEL) {
     assert(cur != NULL, "not NULL-terminated");
     nmethod* next = cur->_oops_do_mark_link;
-    cur->_oops_do_mark_link = NULL;
+    mirror_w(cur)->_oops_do_mark_link = NULL;
     DEBUG_ONLY(cur->verify_oop_relocations());
     NOT_PRODUCT(if (TraceScavenge)  cur->print_on(tty, "oops_do, unmark"));
     cur = next;
@@ -2400,7 +2403,7 @@ void nmethod::copy_scopes_pcs(PcDesc* pcs, int count) {
 
   int size = count * sizeof(PcDesc);
   assert(scopes_pcs_size() >= size, "oob");
-  memcpy(scopes_pcs_begin(), pcs, size);
+  memcpy(mirror_w(scopes_pcs_begin()), pcs, size);
 
   // Adjust the final sentinel downward.
   PcDesc* last_pc = &scopes_pcs_begin()[count-1];
@@ -2408,7 +2411,7 @@ void nmethod::copy_scopes_pcs(PcDesc* pcs, int count) {
   last_pc->set_pc_offset(content_size() + 1);
   for (; last_pc + 1 < scopes_pcs_end(); last_pc += 1) {
     // Fill any rounding gaps with copies of the last record.
-    last_pc[1] = last_pc[0];
+    mirror_w(last_pc)[1] = last_pc[0];
   }
   // The following assert could fail if sizeof(PcDesc) is not
   // an integral multiple of oopSize (the rounding term).
@@ -2419,7 +2422,7 @@ void nmethod::copy_scopes_pcs(PcDesc* pcs, int count) {
 
 void nmethod::copy_scopes_data(u_char* buffer, int size) {
   assert(scopes_data_size() >= size, "oob");
-  memcpy(scopes_data_begin(), buffer, size);
+  memcpy(mirror_w(scopes_data_begin()), buffer, size);
 }
 
 
diff --git a/hotspot/src/share/vm/code/nmethod.hpp b/hotspot/src/share/vm/code/nmethod.hpp
index 0330ca59..07e8357f 100644
--- a/hotspot/src/share/vm/code/nmethod.hpp
+++ b/hotspot/src/share/vm/code/nmethod.hpp
@@ -413,8 +413,8 @@ class nmethod : public CodeBlob {
 
   int total_size        () const;
 
-  void dec_hotness_counter()        { _hotness_counter--; }
-  void set_hotness_counter(int val) { _hotness_counter = val; }
+  void dec_hotness_counter()        { mirror_w(this)->_hotness_counter--; }
+  void set_hotness_counter(int val) { mirror_w_set(_hotness_counter) = val; }
   int  hotness_counter() const      { return _hotness_counter; }
 
   // Containment
@@ -442,7 +442,7 @@ class nmethod : public CodeBlob {
 #if INCLUDE_RTM_OPT
   // rtm state accessing and manipulating
   RTMState  rtm_state() const                     { return _rtm_state; }
-  void set_rtm_state(RTMState state)              { _rtm_state = state; }
+  void set_rtm_state(RTMState state)              { mirror_w_set(_rtm_state) = state; }
 #endif
 
   // Make the nmethod non entrant. The nmethod will continue to be
@@ -457,9 +457,9 @@ class nmethod : public CodeBlob {
 
   // used by jvmti to track if the unload event has been reported
   bool  unload_reported()                         { return _unload_reported; }
-  void  set_unload_reported()                     { _unload_reported = true; }
+  void  set_unload_reported()                     { mirror_w_set(_unload_reported) = true; }
 
-  void set_unloading_next(nmethod* next)          { _unloading_next = next; }
+  void set_unloading_next(nmethod* next)          { mirror_w_set(_unloading_next) = next; }
   nmethod* unloading_next()                       { return _unloading_next; }
 
   static unsigned char global_unloading_clock()   { return _global_unloading_clock; }
@@ -469,7 +469,7 @@ class nmethod : public CodeBlob {
   unsigned char unloading_clock();
 
   bool  is_marked_for_deoptimization() const      { return _marked_for_deoptimization; }
-  void  mark_for_deoptimization()                 { _marked_for_deoptimization = true; }
+  void  mark_for_deoptimization()                 { mirror_w(this)->_marked_for_deoptimization = true; }
 
   void  make_unloaded(BoolObjectClosure* is_alive, oop cause);
 
@@ -478,23 +478,23 @@ class nmethod : public CodeBlob {
   bool has_flushed_dependencies()                 { return _has_flushed_dependencies; }
   void set_has_flushed_dependencies()             {
     assert(!has_flushed_dependencies(), "should only happen once");
-    _has_flushed_dependencies = 1;
+    mirror_w_set(_has_flushed_dependencies) = 1;
   }
 
   bool  is_marked_for_reclamation() const         { return _marked_for_reclamation; }
-  void  mark_for_reclamation()                    { _marked_for_reclamation = 1; }
+  void  mark_for_reclamation()                    { mirror_w(this)->_marked_for_reclamation = 1; }
 
   bool  has_unsafe_access() const                 { return _has_unsafe_access; }
-  void  set_has_unsafe_access(bool z)             { _has_unsafe_access = z; }
+  void  set_has_unsafe_access(bool z)             { mirror_w(this)->_has_unsafe_access = z; }
 
   bool  has_method_handle_invokes() const         { return _has_method_handle_invokes; }
-  void  set_has_method_handle_invokes(bool z)     { _has_method_handle_invokes = z; }
+  void  set_has_method_handle_invokes(bool z)     { mirror_w(this)->_has_method_handle_invokes = z; }
 
   bool  is_lazy_critical_native() const           { return _lazy_critical_native; }
-  void  set_lazy_critical_native(bool z)          { _lazy_critical_native = z; }
+  void  set_lazy_critical_native(bool z)          { mirror_w(this)->_lazy_critical_native = z; }
 
   bool  has_wide_vectors() const                  { return _has_wide_vectors; }
-  void  set_has_wide_vectors(bool z)              { _has_wide_vectors = z; }
+  void  set_has_wide_vectors(bool z)              { mirror_w(this)->_has_wide_vectors = z; }
 
   int   comp_level() const                        { return _comp_level; }
 
@@ -537,29 +537,29 @@ public:
   bool  on_scavenge_root_list() const                  { return (_scavenge_root_state & 1) != 0; }
  protected:
   enum { sl_on_list = 0x01, sl_marked = 0x10 };
-  void  set_on_scavenge_root_list()                    { _scavenge_root_state = sl_on_list; }
-  void  clear_on_scavenge_root_list()                  { _scavenge_root_state = 0; }
+  void  set_on_scavenge_root_list()                    { mirror_w_set(_scavenge_root_state) = sl_on_list; }
+  void  clear_on_scavenge_root_list()                  { mirror_w_set(_scavenge_root_state) = 0; }
   // assertion-checking and pruning logic uses the bits of _scavenge_root_state
 #ifndef PRODUCT
-  void  set_scavenge_root_marked()                     { _scavenge_root_state |= sl_marked; }
-  void  clear_scavenge_root_marked()                   { _scavenge_root_state &= ~sl_marked; }
+  void  set_scavenge_root_marked()                     { mirror_w(this)->_scavenge_root_state |= sl_marked; }
+  void  clear_scavenge_root_marked()                   { mirror_w(this)->_scavenge_root_state &= ~sl_marked; }
   bool  scavenge_root_not_marked()                     { return (_scavenge_root_state &~ sl_on_list) == 0; }
   // N.B. there is no positive marked query, and we only use the not_marked query for asserts.
 #endif //PRODUCT
   nmethod* scavenge_root_link() const                  { return _scavenge_root_link; }
-  void     set_scavenge_root_link(nmethod *n)          { _scavenge_root_link = n; }
+  void     set_scavenge_root_link(nmethod *n)          { mirror_w_set(_scavenge_root_link) = n; }
 
  public:
 
   // Sweeper support
   long  stack_traversal_mark()                    { return _stack_traversal_mark; }
-  void  set_stack_traversal_mark(long l)          { _stack_traversal_mark = l; }
+  void  set_stack_traversal_mark(long l)          { mirror_w_set(_stack_traversal_mark) = l; }
 
   // Exception cache support
   // Note: _exception_cache may be read concurrently. We rely on memory_order_consume here.
   ExceptionCache* exception_cache() const         { return _exception_cache; }
-  void set_exception_cache(ExceptionCache *ec)    { _exception_cache = ec; }
-  void release_set_exception_cache(ExceptionCache *ec) { OrderAccess::release_store_ptr(&_exception_cache, ec); }
+  void set_exception_cache(ExceptionCache *ec)    { mirror_w_set(_exception_cache) = ec; }
+  void release_set_exception_cache(ExceptionCache *ec) { OrderAccess::release_store_ptr(mirror_w(&_exception_cache), ec); }
   address handler_for_exception_and_pc(Handle exception, address pc);
   void add_handler_for_exception_and_pc(Handle exception, address pc, address handler);
   void clean_exception_cache(BoolObjectClosure* is_alive);
@@ -572,7 +572,7 @@ public:
   address  osr_entry() const                      { assert(is_osr_method(), "wrong kind of nmethod"); return _osr_entry_point; }
   void  invalidate_osr_method();
   nmethod* osr_link() const                       { return _osr_link; }
-  void     set_osr_link(nmethod *n)               { _osr_link = n; }
+  void     set_osr_link(nmethod *n)               { mirror_w_set(_osr_link) = n; }
 
   // tells whether frames described by this nmethod can be deoptimized
   // note: native wrappers cannot be deoptimized.
@@ -611,7 +611,7 @@ public:
   bool can_convert_to_zombie();
 
   // Evolution support. We make old (discarded) compiled methods point to new Method*s.
-  void set_method(Method* method) { _method = method; }
+  void set_method(Method* method) { mirror_w_set(_method) = method; }
 
   // GC support
   void do_unloading(BoolObjectClosure* is_alive, bool unloading_occurred);
@@ -677,7 +677,7 @@ public:
   bool is_deopt_mh_entry(address pc) { return pc == deopt_mh_handler_begin(); }
   // Accessor/mutator for the original pc of a frame before a frame was deopted.
   address get_original_pc(const frame* fr) { return *orig_pc_addr(fr); }
-  void    set_original_pc(const frame* fr, address pc) { *orig_pc_addr(fr) = pc; }
+  void    set_original_pc(const frame* fr, address pc) { *mirror_w(orig_pc_addr(fr)) = pc; }
 
   static address get_deopt_original_pc(const frame* fr);
 
diff --git a/hotspot/src/share/vm/code/pcDesc.hpp b/hotspot/src/share/vm/code/pcDesc.hpp
index 83ec9214..5963f347 100644
--- a/hotspot/src/share/vm/code/pcDesc.hpp
+++ b/hotspot/src/share/vm/code/pcDesc.hpp
@@ -26,6 +26,7 @@
 #define SHARE_VM_CODE_PCDESC_HPP
 
 #include "memory/allocation.hpp"
+#include "runtime/os.hpp"
 
 // PcDescs map a physical PC (given as offset from start of nmethod) to
 // the corresponding source scope and byte code index.
@@ -48,7 +49,7 @@ class PcDesc VALUE_OBJ_CLASS_SPEC {
   int _flags;
 
   void set_flag(int mask, bool z) {
-    _flags = z ? (_flags | mask) : (_flags & ~mask);
+    mirror_w_set(_flags) = z ? (_flags | mask) : (_flags & ~mask);
   }
 
  public:
@@ -56,9 +57,9 @@ class PcDesc VALUE_OBJ_CLASS_SPEC {
   int scope_decode_offset() const { return _scope_decode_offset; }
   int obj_decode_offset() const   { return _obj_decode_offset; }
 
-  void set_pc_offset(int x)           { _pc_offset           = x; }
-  void set_scope_decode_offset(int x) { _scope_decode_offset = x; }
-  void set_obj_decode_offset(int x)   { _obj_decode_offset   = x; }
+  void set_pc_offset(int x)           { mirror_w_set(_pc_offset)           = x; }
+  void set_scope_decode_offset(int x) { mirror_w_set(_scope_decode_offset) = x; }
+  void set_obj_decode_offset(int x)   { mirror_w_set(_obj_decode_offset)   = x; }
 
   // Constructor (only used for static in nmethod.cpp)
   // Also used by ScopeDesc::sender()]
diff --git a/hotspot/src/share/vm/code/relocInfo.cpp b/hotspot/src/share/vm/code/relocInfo.cpp
index 343aa043..23f6c4b5 100644
--- a/hotspot/src/share/vm/code/relocInfo.cpp
+++ b/hotspot/src/share/vm/code/relocInfo.cpp
@@ -50,13 +50,13 @@ relocInfo::relocInfo(relocType t, int off, int f) {
 #endif
 
 void relocInfo::initialize(CodeSection* dest, Relocation* reloc) {
-  relocInfo* data = this+1;  // here's where the data might go
+  relocInfo* data = mirror_w(this)+1;  // here's where the data might go
   dest->set_locs_end(data);  // sync end: the next call may read dest.locs_end
   reloc->pack_data_to(dest); // maybe write data into locs, advancing locs_end
-  relocInfo* data_limit = dest->locs_end();
+  relocInfo* data_limit = mirror_w(dest->locs_end());
   if (data_limit > data) {
-    relocInfo suffix = (*this);
-    data_limit = this->finish_prefix((short*) data_limit);
+    relocInfo suffix = (*mirror_w(this));
+    data_limit = mirror_w(this)->finish_prefix((short*) data_limit);
     // Finish up with the suffix.  (Hack note: pack_data_to might edit this.)
     *data_limit = suffix;
     dest->set_locs_end(data_limit+1);
@@ -597,7 +597,7 @@ void trampoline_stub_Relocation::unpack_data() {
 }
 
 void external_word_Relocation::pack_data_to(CodeSection* dest) {
-  short* p = (short*) dest->locs_end();
+  short* p = (short*) mirror_w(dest->locs_end());
   int32_t index = runtime_address_to_index(_target);
 #ifndef _LP64
   p = pack_1_int_to(p, index);
diff --git a/hotspot/src/share/vm/code/stubs.cpp b/hotspot/src/share/vm/code/stubs.cpp
index 9044c5f8..c26e0ae5 100644
--- a/hotspot/src/share/vm/code/stubs.cpp
+++ b/hotspot/src/share/vm/code/stubs.cpp
@@ -72,7 +72,7 @@ StubQueue::StubQueue(StubInterface* stub_interface, int buffer_size,
   _stub_interface  = stub_interface;
   _buffer_size     = blob->content_size();
   _buffer_limit    = blob->content_size();
-  _stub_buffer     = blob->content_begin();
+  _stub_buffer     = mirror_x(blob->content_begin());
   _queue_begin     = 0;
   _queue_end       = 0;
   _number_of_stubs = 0;
diff --git a/hotspot/src/share/vm/code/vtableStubs.cpp b/hotspot/src/share/vm/code/vtableStubs.cpp
index 8d0679be..eb2a22e9 100644
--- a/hotspot/src/share/vm/code/vtableStubs.cpp
+++ b/hotspot/src/share/vm/code/vtableStubs.cpp
@@ -73,7 +73,7 @@ void* VtableStub::operator new(size_t size, int code_size) throw() {
   void* res = _chunk;
   _chunk += real_size;
   align_chunk();
- return res;
+ return mirror_w(res);
 }
 
 
diff --git a/hotspot/src/share/vm/code/vtableStubs.hpp b/hotspot/src/share/vm/code/vtableStubs.hpp
index 04f3e7e3..30fbeb9e 100644
--- a/hotspot/src/share/vm/code/vtableStubs.hpp
+++ b/hotspot/src/share/vm/code/vtableStubs.hpp
@@ -27,6 +27,7 @@
 
 #include "code/vmreg.hpp"
 #include "memory/allocation.hpp"
+#include "runtime/os.hpp"
 
 // A VtableStub holds an individual code stub for a pair (vtable index, #args) for either itables or vtables
 // There's a one-to-one relationship between a VtableStub and such a pair.
@@ -54,7 +55,7 @@ class VtableStub {
   VtableStub* next() const                       { return _next; }
   int index() const                              { return _index; }
   static VMReg receiver_location()               { return _receiver_location; }
-  void set_next(VtableStub* n)                   { _next = n; }
+  void set_next(VtableStub* n)                   { mirror_w_set(_next) = n; }
 
  public:
   address code_begin() const                     { return (address)(this + 1); }
@@ -69,8 +70,8 @@ class VtableStub {
 
  private:
   void set_exception_points(address npe_addr, address ame_addr) {
-    _npe_offset = npe_addr - code_begin();
-    _ame_offset = ame_addr - code_begin();
+    mirror_w_set(_npe_offset) = npe_addr - code_begin();
+    mirror_w_set(_ame_offset) = ame_addr - code_begin();
     assert(is_abstract_method_error(ame_addr),   "offset must be correct");
     assert(is_null_pointer_exception(npe_addr),  "offset must be correct");
     assert(!is_abstract_method_error(npe_addr),  "offset must be correct");
diff --git a/hotspot/src/share/vm/interpreter/interpreter.cpp b/hotspot/src/share/vm/interpreter/interpreter.cpp
index 7ce4bdbb..12b3283b 100644
--- a/hotspot/src/share/vm/interpreter/interpreter.cpp
+++ b/hotspot/src/share/vm/interpreter/interpreter.cpp
@@ -52,8 +52,8 @@
 // Implementation of InterpreterCodelet
 
 void InterpreterCodelet::initialize(const char* description, Bytecodes::Code bytecode) {
-  _description       = description;
-  _bytecode          = bytecode;
+  mirror_w_set(_description) = description;
+  mirror_w_set(_bytecode)    = bytecode;
 }
 
 
diff --git a/hotspot/src/share/vm/interpreter/interpreter.hpp b/hotspot/src/share/vm/interpreter/interpreter.hpp
index ebfb68d3..97ac3584 100644
--- a/hotspot/src/share/vm/interpreter/interpreter.hpp
+++ b/hotspot/src/share/vm/interpreter/interpreter.hpp
@@ -51,7 +51,7 @@ class InterpreterCodelet: public Stub {
  public:
   // Initialization/finalization
   void    initialize(int size,
-                     CodeStrings& strings)       { _size = size;
+                     CodeStrings& strings)       { mirror_w_set(_size) = size;
                                                    DEBUG_ONLY(::new(&_strings) CodeStrings();)
                                                    DEBUG_ONLY(_strings.assign(strings);) }
   void    finalize()                             { ShouldNotCallThis(); }
@@ -117,7 +117,7 @@ class CodeletMark: ResourceMark {
     assert (_clet != NULL, "we checked not enough space already");
 
     // initialize Codelet attributes
-    _clet->initialize(description, bytecode);
+    mirror_w(_clet)->initialize(description, bytecode);
     // create assembler for code generation
     masm  = new InterpreterMacroAssembler(&_cb);
     _masm = &masm;
diff --git a/hotspot/src/share/vm/interpreter/interpreterRuntime.cpp b/hotspot/src/share/vm/interpreter/interpreterRuntime.cpp
index 425ad7f4..fbb42a66 100644
--- a/hotspot/src/share/vm/interpreter/interpreterRuntime.cpp
+++ b/hotspot/src/share/vm/interpreter/interpreterRuntime.cpp
@@ -1158,7 +1158,7 @@ void SignatureHandlerLibrary::initialize() {
 
   BufferBlob* bb = BufferBlob::create("Signature Handler Temp Buffer",
                                       SignatureHandlerLibrary::buffer_size);
-  _buffer = bb->code_begin();
+  _buffer = mirror_x(bb->code_begin());
 
   _fingerprints = new(ResourceObj::C_HEAP, mtCode)GrowableArray<uint64_t>(32, true);
   _handlers     = new(ResourceObj::C_HEAP, mtCode)GrowableArray<address>(32, true);
@@ -1172,7 +1172,7 @@ address SignatureHandlerLibrary::set_handler(CodeBuffer* buffer) {
     handler = set_handler_blob();
   }
   if (handler != NULL) {
-    memcpy(handler, buffer->insts_begin(), insts_size);
+    memcpy(mirror_w(handler), buffer->insts_begin(), insts_size);
     pd_set_handler(handler);
     ICache::invalidate_range(handler, insts_size);
     _handler = handler + insts_size;
diff --git a/hotspot/src/share/vm/memory/heap.cpp b/hotspot/src/share/vm/memory/heap.cpp
index a0fb6034..197c3fe3 100644
--- a/hotspot/src/share/vm/memory/heap.cpp
+++ b/hotspot/src/share/vm/memory/heap.cpp
@@ -193,7 +193,7 @@ void* CodeHeap::allocate(size_t instance_size, bool is_critical) {
     assert(block->length() >= number_of_segments && block->length() < number_of_segments + CodeCacheMinBlockLength, "sanity check");
     assert(!block->free(), "must be marked free");
 #ifdef ASSERT
-    memset((void *)block->allocated_space(), badCodeHeapNewVal, instance_size);
+    memset((void *)mirror_w(block->allocated_space()), badCodeHeapNewVal, instance_size);
 #endif
     return block->allocated_space();
   }
@@ -218,7 +218,7 @@ void* CodeHeap::allocate(size_t instance_size, bool is_critical) {
     b->initialize(number_of_segments);
     _next_segment += number_of_segments;
 #ifdef ASSERT
-    memset((void *)b->allocated_space(), badCodeHeapNewVal, instance_size);
+    memset((void *)mirror_w(b->allocated_space()), badCodeHeapNewVal, instance_size);
 #endif
     return b->allocated_space();
   } else {
@@ -228,12 +228,13 @@ void* CodeHeap::allocate(size_t instance_size, bool is_critical) {
 
 
 void CodeHeap::deallocate(void* p) {
+  p = mirror_x(p);
   assert(p == find_start(p), "illegal deallocation");
   // Find start of HeapBlock
   HeapBlock* b = (((HeapBlock *)p) - 1);
   assert(b->allocated_space() == p, "sanity check");
 #ifdef ASSERT
-  memset((void *)b->allocated_space(),
+  memset((void *)mirror_w(b->allocated_space()),
          badCodeHeapFreeVal,
          segments_to_size(b->length()) - sizeof(HeapBlock));
 #endif
@@ -244,6 +245,7 @@ void CodeHeap::deallocate(void* p) {
 
 
 void* CodeHeap::find_start(void* p) const {
+  p = mirror_x(p);
   if (!contains(p)) {
     return NULL;
   }
diff --git a/hotspot/src/share/vm/memory/heap.hpp b/hotspot/src/share/vm/memory/heap.hpp
index 29e76d7d..b0a74179 100644
--- a/hotspot/src/share/vm/memory/heap.hpp
+++ b/hotspot/src/share/vm/memory/heap.hpp
@@ -26,6 +26,7 @@
 #define SHARE_VM_MEMORY_HEAP_HPP
 
 #include "memory/allocation.hpp"
+#include "runtime/os.hpp"
 #include "runtime/virtualspace.hpp"
 
 // Blocks
@@ -48,15 +49,15 @@ class HeapBlock VALUE_OBJ_CLASS_SPEC {
 
  public:
   // Initialization
-  void initialize(size_t length)                 { _header._length = length; set_used(); }
+  void initialize(size_t length)                 { mirror_w(&_header)->_length = length; set_used(); }
 
   // Accessors
   void* allocated_space() const                  { return (void*)(this + 1); }
   size_t length() const                          { return _header._length; }
 
   // Used/free
-  void set_used()                                { _header._used = true; }
-  void set_free()                                { _header._used = false; }
+  void set_used()                                { mirror_w(&_header)->_used = true; }
+  void set_free()                                { mirror_w(&_header)->_used = false; }
   bool free()                                    { return !_header._used; }
 };
 
@@ -67,14 +68,14 @@ class FreeBlock: public HeapBlock {
 
  public:
   // Initialization
-  void initialize(size_t length)             { HeapBlock::initialize(length); _link= NULL; }
+  void initialize(size_t length)             { HeapBlock::initialize(length); mirror_w_set(_link) = NULL; }
 
   // Merging
-  void set_length(size_t l)                  { _header._length = l; }
+  void set_length(size_t l)                  { mirror_w(&_header)->_length = l; }
 
   // Accessors
   FreeBlock* link() const                    { return _link; }
-  void set_link(FreeBlock* link)             { _link = link; }
+  void set_link(FreeBlock* link)             { mirror_w_set(_link) = link; }
 };
 
 class CodeHeap : public CHeapObj<mtCode> {
@@ -140,7 +141,14 @@ class CodeHeap : public CHeapObj<mtCode> {
   char* high() const                             { return _memory.high(); }
   char* high_boundary() const                    { return _memory.high_boundary(); }
 
-  bool  contains(const void* p) const            { return low_boundary() <= p && p < high(); }
+  bool  contains_internal(const void* p) const   { return low_boundary() <= p && p < high(); }
+  bool  contains(const void* p) const {
+#ifdef __APPLE__
+    return contains_internal(p) || contains_internal((const void*)os::Bsd::mirrored_swap_wx((address)p));
+#else
+    return contains_internal(p);
+#endif
+  }
   void* find_start(void* p) const;              // returns the block containing p or NULL
   size_t alignment_unit() const;                // alignment of any block
   size_t alignment_offset() const;              // offset of first byte of any block, within the enclosing alignment unit
diff --git a/hotspot/src/share/vm/oops/oop.inline.hpp b/hotspot/src/share/vm/oops/oop.inline.hpp
index beec739d..6b8fcc39 100644
--- a/hotspot/src/share/vm/oops/oop.inline.hpp
+++ b/hotspot/src/share/vm/oops/oop.inline.hpp
@@ -276,7 +276,7 @@ inline void oopDesc::store_heap_oop(narrowOop* p, narrowOop v)     { *p = v; }
 inline void oopDesc::encode_store_heap_oop_not_null(narrowOop* p, oop v) {
   *p = encode_heap_oop_not_null(v);
 }
-inline void oopDesc::encode_store_heap_oop_not_null(oop* p, oop v) { *p = v; }
+inline void oopDesc::encode_store_heap_oop_not_null(oop* p, oop v) { *mirror_w(p) = v; }
 
 // Encode and store a heap oop allowing for null.
 inline void oopDesc::encode_store_heap_oop(narrowOop* p, oop v) {
diff --git a/hotspot/src/share/vm/runtime/globals.hpp b/hotspot/src/share/vm/runtime/globals.hpp
index 5caf882e..5adf5f31 100644
--- a/hotspot/src/share/vm/runtime/globals.hpp
+++ b/hotspot/src/share/vm/runtime/globals.hpp
@@ -3411,6 +3411,9 @@ class CommandLineFlags {
   develop_pd(uintx, CodeCacheMinimumUseSpace,                               \
           "Minimum code cache size (in bytes) required to start VM.")       \
                                                                             \
+  product(bool, MirrorMappedCodeCache, false,                               \
+          "Use mirror-mapped code cache for iOS 26")                        \
+                                                                            \
   product_pd(uintx, ReservedCodeCacheSize,                                  \
           "Reserved code cache size (in bytes) - maximum code cache size")  \
                                                                             \
diff --git a/hotspot/src/share/vm/runtime/os.hpp b/hotspot/src/share/vm/runtime/os.hpp
index f91daacf..febb0b72 100644
--- a/hotspot/src/share/vm/runtime/os.hpp
+++ b/hotspot/src/share/vm/runtime/os.hpp
@@ -900,6 +900,12 @@ class os: AllStatic {
 #endif
 #ifdef TARGET_OS_ARCH_bsd_zero
 # include "os_bsd_zero.hpp"
+#endif
+
+#ifndef mirror_w
+# define mirror_w(x) (x)
+# define mirror_x(x) (x)
+# define mirror_w_set(x) (x)
 #endif
 
  public:
